builtin/provisioners/chef-client/ssh_provisioner.go

```diff
@@ -0,0 +1,72 @@
+package chefclient
+
+import (
+	"bytes"
+	"strings"
+
+	"github.com/hashicorp/terraform/communicator"
+	"github.com/hashicorp/terraform/terraform"
+)
+
+func (p *Provisioner) sshInstallChefClient(
+	o terraform.UIOutput,
+	comm communicator.Communicator) error {
+	var installCmd bytes.Buffer
+
+	// Build up a single command based on the given config options
+	installCmd.WriteString("curl")
+	if p.HTTPProxy != "" {
+		installCmd.WriteString(" --proxy " + p.HTTPProxy)
+	}
+	if p.NOProxy != nil {
+		installCmd.WriteString(" --noproxy " + strings.Join(p.NOProxy, ","))
+	}
+	installCmd.WriteString(" -LO https://www.chef.io/chef/install.sh 2>/dev/null &&")
+	if !p.PreventSudo {
+		installCmd.WriteString(" sudo")
+	}
+	installCmd.WriteString(" bash ./install.sh")
+	if p.Version != "" {
+		installCmd.WriteString(" -v " + p.Version)
+	}
+	installCmd.WriteString(" &&")
+	if !p.PreventSudo {
+		installCmd.WriteString(" sudo")
+	}
+	installCmd.WriteString(" rm -f install.sh")
+
+	// Execute the command to install Chef Client
+	return p.runCommand(o, comm, installCmd.String())
+}
+
+func (p *Provisioner) sshCreateConfigFiles(
+	o terraform.UIOutput,
+	comm communicator.Communicator) error {
+	// Make sure the config directory exists
+	if err := p.runCommand(o, comm, "mkdir -p "+linuxConfDir); err != nil {
+		return err
+	}
+
+	// Make sure we have enough rights to upload the files if using sudo
+	if !p.PreventSudo {
+		if err := p.runCommand(o, comm, "chmod 777 "+linuxConfDir); err != nil {
```

Comment:
Yes, but this is OK as this is only done during the uploading of the config files and when you are using sudo. As scp and sudo don't play nice, we need to open the rights for a moment (of our newly created still empty directory) so we can upload the files without any issues.

As you can see [here](https://github.com/svanharmelen/terraform/blob/f-chef-client-provisioner/builtin/provisioners/chef-client/ssh_provisioner.go#L63) the rights are restored right after the copying of the files are finished. And again, if something fails the node will be tainted and recreated. And if nothing fails the result will be `0755` instead of the temporary `0777`.